Enable spell checking and installing dictionaries for QtWebEngine

elshize opened the pull request:
Fixes #700 

Added:
* a module to install/list languages,
* 'spell' and 'spell-languages' settings,
* a command-line script.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/2891)
<!-- Reviewable:end -->

jgkamat commented on qutebrowser/browser/webengine/spell.py:
```diff
@@ -0,0 +1,142 @@
+def get_language_list_file():
+    """Return the path to the file with the list of all available languages."""
+    root_dir = os.path.dirname(os.path.abspath(__file__))
+    # TODO: not sure how to determine the following path 'the right way'
```
Not sure if this is the 'right way', but there's a [basedir](https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/__init__.py) var in the root. I'm not sure if you can dereference it directly or not though.

elshize replied:
Thanks, that definitely sounds more right than what I did. I'll look into it.

NoctuaNivalis commented:
Perhaps it wouldn't be a bad thing to use these dictionaries for [word hints](https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/browser/hints.py#L941) as well? If these dictionaries are just files with valid words seperated by newlines, it should work out of the box when you change the config setting. If not, I'll have a look at it once this is merged.

michal-blippar commented:
These are binary files, not sure what exactly the format is. It's a good idea though, we should investigate.

elshize commented:
These are binary files, not sure what exactly the format is. We should investigate.

NoctuaNivalis commented:
OK, then I'll just make a seperate PR for it later.

elshize commented:
@The-Compiler, just letting you know I'm ready for some feedback. I remember you might still be busy, so take your time.

There are probably some more test I should write, I'd appreciate some pointers.

elshize commented:
Also, I noticed this spell checking doesn't always work. I haven't found anything in the logs. But here's the thing: When I go to Google, and type something weird, it gets underlined as error.

![](https://s3.postimg.org/ahrcvffzn/google_ok.png)

However, it kinda flickers while typing and sometimes it even disappears (and reappears after typing some more):
![](https://s3.postimg.org/4i3lrrv77/google_invisible.png)

Yet, most of the time it doesn't show anything at all, including on Github or Gmail:

![](https://s3.postimg.org/at8t86wfn/github.png)

I'm not sure whether it's because I messed something up, or the spell check is just far from great. Or maybe we treat some forms in a way that overrides this behavior? I'd appreciate some input if anyone has any ideas. Also, is anyone already using the spellcheck and experiencing similar issues?

lahwaacz commented:
Does it happen even on sites without any javascript (or if you turn off javascript)?

elshize commented:
If I disable javascript, it doesn't work at all. If I turn off javascript when I have some stuff unterlined as typos, then from now on it stays underlined in the same position even if I remove the text. So it seems it actually needs javascript to work. Is javascript a part of the web engine component?

When javascript is on, then it behaves like described above.

I tested in Linux (Arch) and now I verified on Mac.

lahwaacz commented:
OK, it's still possible that the javascript code for spell-checking somehow conflicts with the website's code for whatever they need it for, as is the case of GitHub and probably Gmail too. Does it work OK on some "simpler" sites, such as Arch's wiki or bbs?

elshize commented:
Ugh, that's my bad! Since I tested it a long time ago, I've made a typo that caused the dicts not to load, which wasn't reported. And then I tested on Google, which apparently has its own spellcheck. So dumb.

Anyway, I fixed that, added some proper warnings and will push the commit soon. Sorry for the confusion.

The-Compiler commented on misc/requirements/requirements-pylint.txt:
```diff
@@ -4,7 +4,6 @@ astroid==1.5.3
 certifi==2017.7.27.1
 chardet==3.0.4
 github3.py==0.9.6
-idna==2.6
```
All those requirement changes don't seem to belong in this PR.

elshize replied:
How would I go about it if I needed to use a package that's not in the requirements? Say, in my case, I wanted to use `requests` package in my tests.

The-Compiler commented on tests/end2end/features/spell.feature:
```diff
@@ -0,0 +1,38 @@
+# vim: ft=cucumber fileencoding=utf-8 sts=4 sw=4 et:
+
+Feature: Setting spell checking for QtWebEngine
```
Those should be skipped if no dict is available (like when running tests locally)

The-Compiler replied:
Sorry for newer answering this - but you got it right from a quick look!

elshize commented on tests/end2end/features/spell.feature:
```diff
@@ -0,0 +1,38 @@
+# vim: ft=cucumber fileencoding=utf-8 sts=4 sw=4 et:
+
+Feature: Setting spell checking for QtWebEngine
```
What would be the correct way of doing that? Implementing a decorator?

The-Compiler commented:
I'd prefer not doing (remote) network requests in the testsuite at all. I guess a better place to add such a check would be the `scripts/dev/update_3rdparty.py` script which gets run before a release.

elshize commented:
Alright, that makes sense.

The-Compiler commented:
@elshize Any update on this? Also, can you merge the latest master please (and adjust this to work with the new config)?

elshize commented:
It's basically ready, only I have some issues/questions about testing, and didn't want to bother you before you merge the config. I'll merge it as soon as I get a chance. I'll let you know when you can have a look.

The-Compiler commented:
Your PR now contains hundreds of commits I did - did you rebase on `new-config` instead of the latest `master`? :wink:

elshize commented:
I rebased on master alright, just did some weird merge after that unintentionally. Fixed it now. There are still some test I need to fix though.

The-Compiler commented:
The latest failures were my default, restarting the tests with them fixed now :wink:

elshize commented:
I'm having a problem making this work for PyQt5.71. Spell checking is available from 5.8, so I want both `spell` and `spell_languages` options to be available only when run with 5.8 or higher but I'm not sure how to do that.

I tried using `backend` property in `configdata.yml`:
```yaml
backend:
  QtWebKit: false
  QtWebEngine: Qt 5.8
```

But does this do anything? I don't see where it is actually used. In any case, this doesn't work, so I tried setting this in `webenginesettings.py` like this:

```python
if qtutils.version_check('5.8'):
    MAPPINGS['spell'] = DefaultProfileSetter('setSpellCheckEnabled')
    MAPPINGS['spell_languages'] = DictionaryLanguageSetter()
```

This, however, causes the following error:
```
20:25:25 ERROR: Renderer process crashed
/home/qt/work/qt/qtwebengine/src/3rdparty/chromium/sandbox/linux/seccomp-bpf-helpers/sigsys_handlers.cc:**CRASHING**:seccomp-bpf failure in syscall 0281
Received signal 11 SEGV_MAPERR 00000c003119
#0 0x7f9132a02a2e <unknown>
#1 0x7f9132a02deb <unknown>
#2 0x7f912e35eda0 <unknown>
#3 0x7f91331ec048 <unknown>
#4 0x7f913334dc9e <unknown>
#5 0x7f912e35eda0 <unknown>
#6 0x7f912d7a2326 __GI_epoll_pwait
#7 0x7f9133ededf0 <unknown>
#8 0x7f9133edd042 <unknown>
#9 0x7f91329f9d5e <unknown>
#10 0x7f9132a335a8 <unknown>
#11 0x7f9132a1db65 <unknown>
#12 0x7f9132a4bb21 <unknown>
#13 0x7f9132a481c7 <unknown>
#14 0x7f912e35408a start_thread
#15 0x7f912d7a21bf __GI___clone
  r8: 0000000000000002  r9: 00007f911e9e246c r10: 0000000000000000 r11: 0000000000000246
 r12: 00007f911e9e2470 r13: 000000000206d8b0 r14: 00007f911e9e24c0 r15: 000000000206f250
  di: 0000000000000001  si: 00007f9134ae3786  bp: 00007f911e9e24c0  bx: 000000000c003119
  dx: 000000000c000000  ax: 000000000c003000  cx: 00007f912e35d8bb  sp: 00007f911e9e2470
  ip: 00007f91331ec048 efl: 0000000000010202 cgf: 002b000000000033 erf: 0000000000000006
 trp: 000000000000000e msk: 0000000000000000 cr2: 000000000c003119
[end of stack trace]
```

Can I get a little hint as to how to define a setting that's available only starting from some version? I could deal with that in the setter, but then the setting will still be available, only not working...

The-Compiler commented:
You were on the right track - what `backend:` in `configdata.yml` does is mentioning that in the documentation, and showing an error when it's set interactively (e.g. with `:set`) with the wrong backend.

I don't see how that error could be related to the conditionals in `webenginesettings.py`. Can you double-check it's really that causing it? If so, can you please push that version?

elshize commented:
It seems that this error occurs only on pyqt5.71. And you're right, doesn't seem like it's related. I pushed it to see if it fails here as well. I'll look at it later. Also, the first test on Travis fails with `ImportError: libQt5WebEngineWidgets.so.5: cannot open shared object file: No such file or directory`, do you know why that is?

The-Compiler commented on qutebrowser/misc/utilcmds.py:
```diff
@@ -33,6 +33,7 @@
 from PyQt5.QtCore import QUrl
 # so it's available for :debug-pyeval
 from PyQt5.QtWidgets import QApplication  # pylint: disable=unused-import
+from PyQt5.QtWebEngineWidgets import QWebEngineProfile  # pylint: disable=unused-import
```
You can't rely on QtWebEngine being available - but why do you need to import this here at all?

elshize replied:
Right, this is actually something I meant to mention but I forgot. So basically, the reason is as in the comment above for `QApplication`---I wanted it available in `debug-pyeval`. And the reason for that is that I used it in BDD tests to verify that along with the qutebrowser option being set, the `QWebEngineProfile` property is properly set too. See `test_spell_bdd.py`. Now, the question is: can I do it in a cleaner way? This is basically the only way I could think of to test that the spell check is actually working.

The-Compiler replied:
You can do `__import__('QWebEngineProfile').defaultProfile()` etc., but it's kind of a hack... I'll add some comments on those tests.

The-Compiler commented on misc/requirements/requirements-codecov.txt:
```diff
@@ -1,9 +1,9 @@
-urllib3==1.22
+idna==2.5
+requests==2.18.1
+urllib3==1.21.1
```
Looks like you still have changes here compared to the latest `master`

The-Compiler commented on qutebrowser/browser/webengine/langs.tsv:
```diff
@@ -0,0 +1,42 @@
+tg-TG	Tajik (Tajikistan)	tg-TG-5-0.bdic
+tr-TR	Turkish (Turkey)	tr-TR-4-0.bdic
+uk-UA	Ukrainian (Ukraine)	uk-UA-3-0.bdic
+vi-VN	Vietnamese (Viet Nam)	vi-VN-3-0.bdic
```
I don't really like the duplication here to be honest - I think the code inside qutebrowser could use the information from `configdata` instead, and the download script could use the Google API, I'll comment on the script below.

The-Compiler commented on qutebrowser/browser/webengine/spell.py:
```diff
@@ -0,0 +1,150 @@
+
+from PyQt5.QtCore import QLibraryInfo
+
+repository_url = 'https://redirector.gvt1.com/edgedl/chrome/dict/'
```
Instead, I think it should get the file list from [this google API link](https://chromium.googlesource.com/chromium/deps/hunspell_dictionaries.git/+/master/?format=JSON) (giving you [this list](https://chromium.googlesource.com/chromium/deps/hunspell_dictionaries.git/+/master/) as JSON), and then download the files from there (rather than using an internal google host which might change).

The-Compiler commented on qutebrowser/browser/webengine/spell.py:
```diff
@@ -0,0 +1,150 @@
+repository_url = 'https://redirector.gvt1.com/edgedl/chrome/dict/'
+
+
+class Language:
```
qutebrowser recently [introduced](https://github.com/qutebrowser/qutebrowser/commit/3a5241b642da666e4517a32e0eb945254d86a6da) [attrs](http://www.attrs.org/en/stable/), I think this would be a great place to use it.

The-Compiler commented on qutebrowser/browser/webengine/spell.py:
```diff
@@ -0,0 +1,150 @@
+        self.file = file
+
+    @staticmethod
+    def from_array(lang_array):
```
Those are usually called lists in Python, not arrays :wink:

The-Compiler commented on qutebrowser/browser/webengine/spell.py:
```diff
@@ -0,0 +1,150 @@
+                        (code, name, file)
+        """
+        code, name, file = lang_tuple
+        return Language(code, name, file)
```
I don't think this (and `from_array`) is needed, as the calling code can just do `Language(*a_tuple)` or `Language(*a_list)`.

The-Compiler commented on qutebrowser/browser/webengine/spell.py:
```diff
@@ -0,0 +1,150 @@
+
+import os
+from urllib.parse import urljoin
+from urllib.request import urlretrieve
```
Please don't import functions from modules - instead just import `urllib.parse` and `urllib.request`, then use `urllib.parse.urljoin` and `urllib.request.urlretrieve`.

The-Compiler commented on qutebrowser/browser/webengine/spell.py:
```diff
@@ -0,0 +1,150 @@
+    return filtered_languages
+
+
+def download_dictionary(url, dest):
```
This and `install` look like they'd fit better in `scripts/install_dict.py` instead of here, since they're intended for that script and not usage inside qutebrowser.

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -127,6 +127,29 @@ def _set(self, value, settings=None):
+            raise ValueError("'settings' may not be set with "
+                             "DictionaryLanguageSetter!")
+        installed_langs = dict([(lang.code, lang.file)
+                                for lang in get_installed_languages()])
```
`{lang.code: lang.file for lang in get_installed_languages()}`

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -299,6 +322,7 @@ def shutdown():
 
     'scrolling.smooth':
         Attribute(QWebEngineSettings.ScrollAnimatorEnabled),
+
```
unrelated change

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -975,6 +975,71 @@ scrolling.smooth:
+    QtWebKit: false
+    QtWebEngine: Qt 5.8
+
+spell_languages:
```
Maybe that should be `spell.enabled` and `spell.languages`? I also wonder whether it'd be better to call it `spellcheck` instead...

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -127,6 +127,29 @@ def _set(self, value, settings=None):
+        lang_files = []
+        for lang_code in value:
+            if lang_code in installed_langs:
+                lang_files.append(installed_langs[lang_code][:-5])
```
Use `os.path.splitext(...)[0]` instead - however, maybe it'd be nicer for `installed_langs` to map from `lang.code` to `lang`, and make `Language` have some kind of `filename` attribute?

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -975,6 +975,71 @@ scrolling.smooth:
+        - vi-VN: Vietnamese (Viet Nam)
+    none_ok: true
+  default:
+  desc: Spell checking languages.
```
Should probably have some documentation on how to install them.

The-Compiler commented on tests/end2end/conftest.py:
```diff
@@ -29,6 +29,8 @@
 import os.path
 import operator
 
+from qutebrowser.browser.webengine.spell import get_installed_languages
```
`from qutebrowser.browser.webengine import spell` and use `spell.get_installed_languages()`

The-Compiler commented on tests/end2end/conftest.py:
```diff
@@ -116,14 +118,36 @@ def _get_backend_tag(tag):
 
+def _get_dictionary_tag(tag):
+    """Handle tags like must_have_dict=en-US for BDD tests."""
+    version_re = re.compile(r"""
```
`dict_re` or `tag_re` or so

The-Compiler commented on tests/end2end/conftest.py:
```diff
@@ -116,14 +118,36 @@ def _get_backend_tag(tag):
+
+    match = version_re.match(tag)
+    if not match:
+        #return pytest.mark.skip
```
You can probably delete this

The-Compiler commented on tests/end2end/conftest.py:
```diff
@@ -29,6 +29,8 @@
 import os.path
```
See my comments on `spell.feature` before looking at those please :wink:

The-Compiler commented on tests/end2end/features/spell.feature:
```diff
@@ -0,0 +1,38 @@
+# vim: ft=cucumber fileencoding=utf-8 sts=4 sw=4 et:
```
These tests look quite hacky (I'd like to avoid `:debug-pyeval` in feature tests), and I also don't see a reason why they need to be end2end tests (they don't really test for proper integration with QtWebEngine/Chromium, but they can't).

I think it'd be better if those were unit tests for `webenginesettings.py` instead. Then it's much easier to check the profile, and they also are much faster and more reliable that way.

The-Compiler commented on tests/unit/browser/webengine/test_spell.py:
```diff
@@ -0,0 +1,95 @@
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+
+from os.path import basename, join
```
Like above - please import modules, not things from modules (I find it hard to keep track what's from where otherwise)

The-Compiler commented on tests/unit/browser/webengine/test_spell.py:
```diff
@@ -0,0 +1,95 @@
+
+
+def test_get_installed_languages_empty(tmpdir, mocker):
+    mocker.patch('qutebrowser.browser.webengine.spell.get_dictionary_dir',
```
For tests where you need to patch something but don't need the unittest.mock `assert_*` methods, please use [monkeypatch](https://docs.pytest.org/en/latest/monkeypatch.html)

The-Compiler commented on tests/unit/browser/webengine/test_spell.py:
```diff
@@ -0,0 +1,95 @@
+    mocker.patch('qutebrowser.browser.webengine.spell.get_dictionary_dir',
+                 lambda: str(tmpdir))
+    for lang in LANGUAGE_LIST:
+        open(join(str(tmpdir), lang.file), 'w', encoding='UTF-8').close()
```
`(tmpdir / lang.file).ensure()` (see the [py.path docs](http://py.readthedocs.io/en/latest/path.html))

The-Compiler commented on tests/unit/browser/webengine/test_spell.py:
```diff
@@ -0,0 +1,95 @@
+    for actual, expected in zip(spell.get_installed_languages(),
+                                LANGUAGE_LIST):
+        assert (actual.code, actual.name, actual.file) ==\
+               (expected.code, expected.name, expected.file)
```
After switching to `attrs`, you can just do `assert actual == expected`

The-Compiler commented on tests/unit/browser/webengine/test_spell.py:
```diff
@@ -0,0 +1,95 @@
+    assert len(language_list) == 42
+    first_lang = language_list[0]
+    assert (first_lang.code, first_lang.name, first_lang.file) ==\
+           (AFRIKAANS.code, AFRIKAANS.name, AFRIKAANS.file)
```
Same as above.

The-Compiler commented on tests/unit/browser/webengine/test_spell.py:
```diff
@@ -0,0 +1,95 @@
+    mocker.patch('qutebrowser.browser.webengine.spell.get_dictionary_dir',
+                 lambda: str(tmpdir))
+    mocker.patch('qutebrowser.browser.webengine.spell.download_dictionary',
+                 lambda url, dest: open(dest, 'w', encoding='UTF-8').close())
```
`_url` because it's unused, and maybe `import py.path` and use `py.path(dest).ensure()` here.

The-Compiler commented on tests/unit/browser/webengine/test_spell.py:
```diff
@@ -0,0 +1,95 @@
+    mocker.patch('qutebrowser.browser.webengine.spell.download_dictionary',
+                 lambda url, dest: open(dest, 'w', encoding='UTF-8').close())
+    spell.install(LANGUAGE_LIST)
+    installed_files = [basename(str(file)) for file in tmpdir.listdir()]
```
`file.basename`

The-Compiler commented on tests/unit/browser/webengine/test_spell.py:
```diff
@@ -0,0 +1,95 @@
+    assert sorted(installed_files) == sorted(expected_files)
+
+
+# TODO: move somewhere to be checked before a release
```
`update_3rdparty.py` / `build_release.py` would probably be the right place

elshize commented:
Thanks for the comments, hopefully I'll find some time this weekend to make the changes.

elshize commented:
Unfortunately, I don't have as much time as I hoped for, I'll have to come back to this next week. Just letting you know.

elshize commented:
I fixed things according to your comments, I also rearranged and cleaned things up a bit. Let me know what else you find.

There are still, however, two outstanding issues:
* I'm not sure how to write those unit tests for webenginesettings you mentioned in your comments. Any hints or existing code examples?
* In `language_list_from_api`, when I retrieve the list of dictionaries, the JSON file has some garbage before the first curly bracket: `)]}'`. I can't figure out what it is and why it's there. Any ideas? I put a workaround there for the time being.

The-Compiler commented:
As for the tests: Since we don't really have any tests for `webenginesettings.py` yet anyways, I'm also okay with not really testing those bits currently.

I don't think it makes sense to have dedicated tests for `spellcheck.enabled` (that'd be tested fine if we had tests for `DefaultProfileSetter`) , but some tests for the language list probably make sense, and I'm okay with leaving those as end2end tests for now (in the long run, I'd like to convert more end2end tests to unit tests in general, as they turned out to be a somewhat bad idea).

I'll leave some more inline comments.

---

Looks like that JSON prefix is intentional:

https://github.com/google/gitiles/issues/22
https://github.com/google/gitiles/issues/82

I suggest stripping out 5 bytes as recommended there (i.e. including the newline), and adding a comment with the link above.

The-Compiler commented on scripts/dev/update_3rdparty.py:
```diff
@@ -110,14 +111,34 @@ def update_ace():
+    from qutebrowser.config import configdata
+    configdata.init()
+    for lang in install_dict.available_languages():
+        sys.stdout.write('Testing dictionary {}... '.format(lang.code))
```
I'd prefer `print(..., end='')` here - I'm not sure whether it makes any difference, but it's a common pattern for this kind of thing.

The-Compiler commented on scripts/dev/update_3rdparty.py:
```diff
@@ -110,14 +111,34 @@ def update_ace():
+
+
+def run(ace=False, pdfjs=True, fancy_dmg=False, pdfjs_version=None,
+        dicts=None):
```
Why `None` and not `False`?

The-Compiler commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,178 @@
+
+def print_list(languages):
+    for lang in languages:
+        print('{1}\t{0}'.format(lang.name, lang.code))
```
You can just swap `lang.code` and `lang.name` now (and use `{}` instead of numbered references)

NoctuaNivalis replied:
Or `print(lang.code, lang.name, sep='\t')`?

The-Compiler replied:
TIL about `sep=` for `print`... Nice!

The-Compiler commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,178 @@
+def download_dictionary(url, dest):
+    """Download a decoded dictionary file."""
+    response = urllib.request.urlopen(url)
+    decoded = base64.decodebytes(response.read())
```
I really wish there was a way to just download a file from Google's repository browser without involving base64, but I couldn't find one... weird.

The-Compiler commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,178 @@
+        selected: a list of keys to select
+        by: a function returning the selection key (code by default)
+        fail_on_unknown: whether to raise an error if there is an unknown
+                         key in selected
```
`by` and `fail_on_unknown` seem to be gone now

The-Compiler commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,178 @@
+        try:
+            install(filter_languages(languages, args.languages))
+        except ValueError as e:
+            print(e)
```
This seems like something likely to bite someone in the future, if something else accidentally raises a `ValueError`. I think it'd be better to define a custom `Error` exception in this module, and raise/handle that.

The-Compiler commented on tests/end2end/features/spell.feature:
```diff
@@ -0,0 +1,30 @@
+    @qtwebkit_skip @qt>=5.8
+    Scenario: Set an invalid language
+        When I run :set spellcheck.languages ['invalid-language'] (invalid command)
+        Then the error "set: Invalid value 'invalid-language' *" should be shown
```
The tests above only test the normal setting machinery and nothing specific to your changes really - so I'd drop them.

The-Compiler commented on tests/end2end/features/spell.feature:
```diff
@@ -0,0 +1,30 @@
+    @qtwebkit_skip @qt>=5.8 @must_have_dict=en-US
+    Scenario: Set valid and installed language
+        When I run :set spellcheck.languages ["en-US"]
+        Then the option spellcheck.languages should be set to ["en-US"]
```
These two tests seem useful, but I think they can go to `misc.features` instead of having their own file.

The-Compiler commented on tests/unit/browser/webengine/test_spell.py:
```diff
@@ -0,0 +1,40 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
```
You can now add the `spell.py` file and this to `PERFECT_FILES` in `scripts/dev/check_coverage.py`.

elshize commented on qutebrowser/config/configdata.yml:
```diff
@@ -975,6 +975,75 @@ scrolling.smooth:
 
+## spell
+
+spellcheck.enabled:
```
Well, technically, setting an empty list of languages effectively turns off spellchecking. So we could just leave it enabled. I guess this setting is more about temporarily turning spellcheck off and not having to erase the list of your preferred languages and later type it in again. I'm not sure how useful it actually is, though. It's your call.

The-Compiler replied:
Ah, if that works without warnings from Qt, I'd say let's remove it. Users could always bind a key to something like `:config-cycle spellcheck.languages [en-US] []` I think

elshize replied:
I wasn't even aware of `config-cycle`, it sounds useful ;) But yeah, I'm not experiencing any warnings whatsoever so we might as well just drop it.

elshize commented:
A suggestion: you might consider adding `*.un~` to `.gitignore`.

The-Compiler commented:
What's `*.un~` and where's it from?

elshize commented:
It's vim's undo file (persists undo history).

elshize commented:
Any idea why `:window-only` test on OSX=sierra fails? I can't reproduce this.

The-Compiler commented:
Looks like a flaky test - don't worry about it. I hopefully fixed it in d3d644d6a5aefe62d6a4899aa9fdf8efaf887a78 and did a re-run.

As for `.gitignore`: I'd rather not have editor-specific files there, I should probably remove `*.swp` as well. You might want to set `undodir` to have them in a central location, and/or configure a `~/.gitignore` to ignore any vim files.

elshize commented:
Okay, then, let me know if there's anything else or if I missed anything.

As for `.gitignore`, you already have vim modelines in all files, so it only felt natural to me ;) But yeah, the tip about `undodir` is good, I'll try doing that.

The-Compiler commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,185 @@
+import re
+import urllib.parse
+import urllib.request
+import attr
```
Please add a newline above `import attr` to group imports according to stdlib/third-party/local (see [pep8](http://pep8.org/#imports))

The-Compiler commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,185 @@
+import urllib.parse
+import urllib.request
+import attr
+
```
Add a `sys.path.insert(0, os.path.join(os.path.dirname(__file__), os.pardir))` here too so this can be invoked like `python3 scripts/install_dict.py` as well.

The-Compiler commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,185 @@
+
+
+class InvalidLanguageError(Exception):
+    """Raised when requested invalid languages."""
```
Please add an empty line above the docstring - [pep257](https://www.python.org/dev/peps/pep-0257/) used to recommend that but changed, however qutebrowser didn't adjust yet, so I think it's best to be consistent.

The-Compiler replied:
requested -> requesting?

The-Compiler commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,185 @@
+
+@attr.s
+class Language:
+    """Dictionary language specs."""
```
Same as above.

The-Compiler commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,185 @@
+
+    def __init__(self, invalid_langs):
+        msg = 'invalid languages: {}'.format(', '.join(invalid_langs))
+        super(InvalidLanguageError, self).__init__(msg)
```
Use `super()` as we don't support Python 2

The-Compiler commented on tests/end2end/features/misc.feature:
```diff
@@ -540,3 +540,15 @@ Feature: Various utility commands.
+    @qtwebkit_skip @qt>=5.8 @must_have_dict=en-US
+    Scenario: Set valid and installed language
+        When I run :set spellcheck.languages ["en-US"]
+        Then the option spellcheck.languages should be set to ["en-US"]
```
Please add a newline (but not an empty line) at the end of the file.

The-Compiler commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,185 @@
+        try:
+            install_lang(lang)
+        except PermissionError as e:
+            print(e)
```
Maybe `sys.exit(1)` here? Doesn't really make sense to go on if it didn't work for one of the files.

The-Compiler commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,190 @@
+    print('Installing {}: {}'.format(lang.code, lang.name))
+    lang_url = urllib.parse.urljoin(API_URL, lang.file_path, '?format=TEXT')
+    if not os.path.isdir(spell.dictionary_dir()):
+        warn_msg = 'WARN: {} does not exist, creating the directory'
```
No need for the `WARN: ` prefix here I think - it's not like something bad happened :wink:

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -127,6 +127,27 @@ def _set(self, value, settings=None):
+    def _find_installed(self, code):
+        installed_file = spell.installed_file(code)
+        if not installed_file:
+            message.warning('Language {} is not installed.'.format(code))
```
Maybe add a ` - see scripts/install_dict.py in qutebrowser's sources` or so?

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -309,6 +330,10 @@ def shutdown():
 
 
+if qtutils.version_check('5.8'):
+    MAPPINGS['spellcheck.languages'] = DictionaryLanguageSetter()
```
You now never call `setSpellCheckEnabled(True)` on the profiles, you should probably do so on Qt 5.8 in `_init_profiles` (but like discussed, without any setting).

Curiosly, on e.g. https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_textarea I still see the URL underlined when I activate the field, but as soon as I start typing it's gone when not enabling spell checking on the profile (I did so via `:debug-console`).

The-Compiler commented:
Merged - thanks a lot for the nice work and the patience!

We might want to revisit this when Qt 5.10 is released (probably in December) - then with `QTWEBENGINE_DICTIONARIES_PATH`, qutebrowser could download this dictionaries itself without a script and without requiring root permissions. Is this something you can see yourself working on, @elshize?

lahwaacz commented on scripts/install_dict.py:
```diff
@@ -0,0 +1,190 @@
+def install_lang(lang):
+    """Install a single lang given by the argument."""
+    print('Installing {}: {}'.format(lang.code, lang.name))
+    lang_url = urllib.parse.urljoin(API_URL, lang.file_path, '?format=TEXT')
```
This does not do what you think, because the signature of the [urllib.parse.urljoin](https://docs.python.org/3.6/library/urllib.parse.html?highlight=urljoin#urllib.parse.urljoin) function is:

```python
urllib.parse.urljoin(base, url, allow_fragments=True)
```

The-Compiler replied:
Good catch, thanks! I fixed this in 6c300f41dd228a1912dc9617632d3cfd70c696b5.

lahwaacz replied:
Nothing spectacular, spell-checking just didn't work no matter what and all dictionaries had only 14 bytes or so...

The-Compiler replied:
Hmm, that's odd - I also have a 14-byte `/usr/share/qt/qtwebengine_dictionaries/en-US-7-1.bdic` but spellchecking did work for me somehow.

elshize commented:
Yeah, I'd be happy to work on that.

The-Compiler commented:
Cool! I mentioned you in #2939 for now, I'll probably open a dedicated issue for it when Qt 5.10 is out.

elshize commented:
Great. I hope to work on other things as well when I find some extra time.

The-Compiler commented:
> I'm not sure how to write those unit tests for webenginesettings you mentioned in your comments. Any hints or existing code examples?

Quick update: After some `webenginesettings.py` refactoring and [fixing a spellcheck bug](https://github.com/qutebrowser/qutebrowser/commit/824825e67da44a0eab49182db6efd5423a8aa8b0) I introduced, I took another look at the spell end2end tests and was able to [easily convert them to unittests](https://github.com/qutebrowser/qutebrowser/commit/f3aaa1084a382296019ee79b479286d45a45a8e5) now :smile:
